src/modules/editor/SoundEditor/SoundEditor.tsx

```diff
@@ -1,70 +1,266 @@
+    this.provider = provider;
+    this.ydoc = ydoc;
+  }
+
+  sendData(data: string): void {
+    console.log("SoundEditor sendData", data);
   }
 
-  public init(doc: Doc, provider: WebrtcProvider): void {
+  loadData(data: string): void {
+    console.log("SoundEditor loadData", data);
   }
 
+  handleCellClick(endRow: number, endCol: number): void {
+
+    if (this._isMouseDown) {
+      if (this._startPosition[0] == -1) {
+        this._startPosition[0] = endRow;
+        this._startPosition[1] = endCol;
+      }
+      return;
+    }
+    if (endRow == this._startPosition[0]) {
+
+      for (let i = Math.min(this._startPosition[1], endCol); i <= Math.max(this._startPosition[1], endCol); i++) {
+        const row = this._startPosition[0];
+        const col = i;
+        const cellKey = `${row}-${col}`;
+
+        this.setState((prevState) => {
+          const newActiveCells = new Set(prevState.activeCells);
+          if (newActiveCells.has(cellKey)) {
+            newActiveCells.delete(cellKey);
+          } else {
+            newActiveCells.add(cellKey);
+          }
+          return { activeCells: newActiveCells };
+        });
+        this.setState({ startPosition: [-1, -1] });
+      }
+      this.state.currentMusic.setNote(this._startPosition[1], this._startPosition[0], Math.max(1, Math.abs(this._startPosition[1] - endCol)), this.state.currentInstrument);
+    }
+
+  }
+
+  clearMusic() {
+    this.setState({ activeCells: new Set<string>() });
+    this.state.currentMusic = new Music();
+  }
+
+  getInstrumentButtons() {
+    return Array.from(instruments.keys()).map((instrument) => (
+      <MusicEditorButton
+        className={`flex-item-grow ${this.state.currentInstrument === instrument ? "selected" : ""}`}
+        key={instrument}
+        onClick={() => {
+          this.setState({ currentInstrument: instrument });
+        }}
+      >
+        {instruments.get(instrument)}
+      </MusicEditorButton>
+    ));
+  }
+
+  saveMusic() {
+    const musicData = this.state.currentMusic.toJson();
+    console.log("Music data to save:", musicData);
+    this._musics[this.state.selectedMusicIndex] = (Music.fromJson(musicData));
+  }
+
+  loadStateFromMusic(id: number) {
+    const music = this._musics[id];
+    this.setState({ selectedMusicIndex: id });
+    this.setState({ currentMusic: music });
+    this.setState({ activeCells: new Set<string>() });
+    for (let i = 0; i < music.notes.length; i++) {
+      for (let j = 0; j < music.notes[i].length; j++) {
+        if (music.notes[i][j].note != "Nan") {
+          const cellKey = `${j}-${i}`;
+          this.setState((prevState) => {
+            const newActiveCells = new Set(prevState.activeCells);
+            newActiveCells.add(cellKey);
+            return { activeCells: newActiveCells };
+          });
+        }
+      }
+    }
+  }
+
+  componentDidMount() {
+    window.addEventListener("mouseup", this.handleMouseUp);
```

Comment:
Please avoid using pure-JS mechanisms for events. Use the corresponding React `onMouseUp` and `onMouseDown` events to avoid UB.